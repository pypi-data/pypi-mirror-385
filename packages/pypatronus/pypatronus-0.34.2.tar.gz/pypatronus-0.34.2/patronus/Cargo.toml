[package]
name = "patronus"
version.workspace = true
description = "Hardware bug-finding toolkit."
homepage = "https://kevinlaeufer.com"
keywords = ["RTL", "btor", "model-checking", "SMT", "bit-vector"]
edition.workspace = true
authors.workspace = true
repository.workspace = true
readme = "Readme.md"
license.workspace = true
rust-version.workspace = true

[dependencies]
indexmap = "2.7.0"
codespan-reporting = "0.11.1"
lazy_static = "1.4.0"
smallvec = { version = "1.x", features = ["union"] }
rand = { version = "0.9", features = ["small_rng"] }
boolean_expression.workspace = true
regex.workspace = true
baa.workspace = true
rustc-hash.workspace = true
thiserror.workspace = true
fst-writer = "0.3.1"

[dev-dependencies]
insta = { version = "1.x", features = ["yaml"] }

