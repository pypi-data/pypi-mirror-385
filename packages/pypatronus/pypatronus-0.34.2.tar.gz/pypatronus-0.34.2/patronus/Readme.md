# Patronus


[![Crates.io Version](https://img.shields.io/crates/v/patronus)](https://crates.io/crates/patronus)
[![docs.rs](https://img.shields.io/docsrs/patronus)](https://docs.rs/patronus)
[![GitHub License](https://img.shields.io/github/license/ekiwi/patronus)](LICENSE)



### TODO

Some things we will hopefully get to one day.

- simulator
  - JIT based implementation
  - better debugging, add option to print expressions with trace
  - waveform generation
  - quickly update only parts of the circuit


#### API Changes

- `patronus::btor2::parse_file` should take in a `Context` instead of producing one
- `Context` should use `RefCell` to allow expressions to be built with a immutable reference