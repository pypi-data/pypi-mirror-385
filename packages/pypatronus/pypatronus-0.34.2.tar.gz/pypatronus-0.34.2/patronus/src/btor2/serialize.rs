// Copyright 2023-2024 The Regents of the University of California
// Copyright 2024 Cornell University
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@cornell.edu>
use crate::expr::*;
use crate::system::TransitionSystem;
use std::io::Write;

pub fn serialize(
    ctx: &Context,
    writer: &mut impl Write,
    sys: &TransitionSystem,
) -> std::io::Result<()> {
    Serializer::new(ctx, writer).serialize_sys(sys)
}

const VERSION: Option<&str> = option_env!("CARGO_PKG_VERSION");

pub fn serialize_to_str(ctx: &Context, sys: &TransitionSystem) -> String {
    let mut buf = Vec::new();
    serialize(ctx, &mut buf, sys).expect("Failed to write to string!");
    String::from_utf8(buf).expect("Failed to read string we wrote!")
}

struct Serializer<'a, W: Write> {
    #[allow(dead_code)] // TODO: implement serializer which will use the ctx!
    ctx: &'a Context,
    writer: &'a mut W,
}

impl<'a, W: Write> Serializer<'a, W> {
    fn new(ctx: &'a Context, writer: &'a mut W) -> Self {
        Serializer { ctx, writer }
    }

    fn serialize_sys(&mut self, sys: &TransitionSystem) -> std::io::Result<()> {
        writeln!(
            self.writer,
            "; btor2 description of `{}` generated by patron {}",
            sys.name,
            VERSION.unwrap_or_default()
        )?;

        writeln!(self.writer, "; TODO: implement the btor2 serialization!")?;

        Ok(())
    }
}
