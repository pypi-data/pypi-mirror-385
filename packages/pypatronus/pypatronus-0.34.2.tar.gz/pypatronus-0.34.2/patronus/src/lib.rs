// Copyright 2023 The Regents of the University of California
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@berkeley.edu>
#[macro_use]
extern crate lazy_static;
extern crate core;

pub mod btor2;
pub mod expr;
pub mod mc;
pub mod sim;
pub mod smt;
pub mod system;
