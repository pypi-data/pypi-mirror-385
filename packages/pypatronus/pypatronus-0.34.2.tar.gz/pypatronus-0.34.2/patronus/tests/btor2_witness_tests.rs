// Copyright 2023 The Regents of the University of California
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@berkeley.edu>

use patronus::btor2;

#[test]
fn test_no_state_witness() {
    let wit = btor2::parse_witness(&mut NO_STATE.as_bytes()).unwrap();
    let serialized = btor2::witness_to_string(&wit);
    insta::assert_snapshot!(serialized);
}

#[test]
fn test_fsm_witness() {
    let wit = btor2::parse_witness(&mut FSM.as_bytes()).unwrap();
    let serialized = btor2::witness_to_string(&wit);
    insta::assert_snapshot!(serialized);
}

#[test]
fn test_const_array_example_witness() {
    let wit = btor2::parse_witness(&mut CONST_ARRAY_EXAMPLE.as_bytes()).unwrap();
    let serialized = btor2::witness_to_string(&wit);
    insta::assert_snapshot!(serialized);
}

#[test]
fn test_const_array_example_witness_2() {
    let wit = btor2::parse_witness(&mut CONST_ARRAY_EXAMPLE_2.as_bytes()).unwrap();
    let serialized = btor2::witness_to_string(&wit);
    assert_eq!(serialized.trim(), CONST_ARRAY_EXAMPLE_2.trim());
}

#[test]
fn test_multiple_witnesses() {
    let wit = btor2::parse_witness(&mut MULTIPLE.as_bytes()).unwrap();
    let serialized = btor2::witness_to_string(&wit);
    insta::assert_snapshot!(serialized);
    let _wits = btor2::parse_witnesses(&mut MULTIPLE.as_bytes(), 30).unwrap();
}

const NO_STATE: &str = r#"
sat
b0
@0
0 0 reset@0
1 11111011 a@0
2 00000101 b@0
.

"#;

const FSM: &str = r#"
sat
b0
#0
0 00 state#0
@0
0 1 reset@0
1 1 in@0
@1
0 0 reset@1
1 1 in@1
.

"#;

/// Witness obtained by running `btormc const_array_example.btor`
const CONST_ARRAY_EXAMPLE: &str = r#"
sat
b0
#0
0 11111 addr#0
1 11111111111111111111111111111111 data#0
2 [11111] 11111111111111111111111111111110 mem@0
4 11111 a#0
@0
.
"#;

// modified version of the original, to test multiple array updates
const CONST_ARRAY_EXAMPLE_2: &str = r#"
sat
b0
#0
0 11111 addr#0
1 11111111111111111111111111111111 data#0
2 [11110] 11111111111111111111111111111110 mem#0
2 [11111] 11111111111111111111111111111110 mem#0
4 11111 a#0
@0
.
"#;

const MULTIPLE: &str = r#"
sat
b0
#0
0 0 state0#0
2 0 state2#0
3 0 state3#0
4 0 state4#0
6 1 state6#0
7 0 state7#0
8 0 state8#0
9 0 state9#0
10 0 state10#0
11 0 state11#0
12 1 state12#0
@0
0 0 clock@0
1 1 in@0
2 1 reset@0
#1
8 0 state8#1
9 0 state9#1
10 0 state10#1
11 0 state11#1
12 0 state12#1
@1
0 0 clock@1
1 0 in@1
2 0 reset@1
#2
8 0 state8#2
9 0 state9#2
10 0 state10#2
11 0 state11#2
12 0 state12#2
@2
0 0 clock@2
1 0 in@2
2 0 reset@2
.
sat
b1
#0
0 0 state0#0
2 0 state2#0
3 0 state3#0
4 0 state4#0
6 1 state6#0
7 0 state7#0
8 0 state8#0
9 0 state9#0
10 0 state10#0
11 0 state11#0
12 1 state12#0
@0
0 0 clock@0
1 1 in@0
2 1 reset@0
#1
8 0 state8#1
9 0 state9#1
10 0 state10#1
11 0 state11#1
12 0 state12#1
@1
0 0 clock@1
1 1 in@1
2 0 reset@1
#2
8 0 state8#2
9 0 state9#2
10 0 state10#2
11 0 state11#2
12 0 state12#2
@2
0 0 clock@2
1 0 in@2
2 0 reset@2
#3
8 0 state8#3
9 0 state9#3
10 0 state10#3
11 0 state11#3
12 0 state12#3
@3
0 0 clock@3
1 0 in@3
2 0 reset@3
.
sat
b2
#0
0 0 state0#0
2 0 state2#0
3 0 state3#0
4 0 state4#0
6 1 state6#0
7 0 state7#0
8 0 state8#0
9 0 state9#0
10 0 state10#0
11 0 state11#0
12 1 state12#0
@0
0 0 clock@0
1 1 in@0
2 1 reset@0
#1
8 0 state8#1
9 0 state9#1
10 0 state10#1
11 0 state11#1
12 0 state12#1
@1
0 0 clock@1
1 1 in@1
2 0 reset@1
#2
8 0 state8#2
9 0 state9#2
10 0 state10#2
11 0 state11#2
12 1 state12#2
@2
0 0 clock@2
1 1 in@2
2 0 reset@2
#3
8 0 state8#3
9 0 state9#3
10 0 state10#3
11 0 state11#3
12 0 state12#3
@3
0 0 clock@3
1 0 in@3
2 0 reset@3
#4
8 0 state8#4
9 0 state9#4
10 0 state10#4
11 0 state11#4
12 0 state12#4
@4
0 0 clock@4
1 0 in@4
2 0 reset@4
.

"#;
