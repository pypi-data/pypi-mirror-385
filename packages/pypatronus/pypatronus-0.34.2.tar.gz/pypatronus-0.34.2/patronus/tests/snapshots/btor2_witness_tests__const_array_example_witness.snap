---
source: tests/btor2_witness_tests.rs
expression: serialized
---
sat
b0
#0
0 11111 addr#0
1 11111111111111111111111111111111 data#0
2 [11111] 11111111111111111111111111111110 mem#0
4 11111 a#0
@0
.

